pkg/queue/metrics.go

```diff
@@ -24,70 +16,137 @@ type TaskQueueMetricsType struct {
-
-// SetupSchedulerMetrics must be called before any other call to the metric subsystem happens
-func SetupSchedulerMetrics(namespace string) {
-	oneSchedulerMetricsSetup.Do(func() {
-		SchedulerMetrics = SchedulerMetricsType{
-			Labels: queueMetricLabels,
-			ScheduleCounter: promauto.NewCounterVec(
-				prometheus.CounterOpts{
-					Namespace: namespace,
-					Subsystem: "scheduler",
-					Name:      "task",
-					Help:      "count of tasks that have been scheduled",
-				},
-				taskMetricLabels,
-			),
-			ErrorCounter: promauto.NewCounterVec(
-				prometheus.CounterOpts{
-					Namespace: namespace,
-					Subsystem: "scheduler",
-					Name:      "error",
-					Help:      "count of errors while scheduling",
-				},
-				taskMetricLabels,
-			),
-		}
-	})
+const (
+	instanceKey = "instance"
+	serviceKey  = "service"
+)
+
+var (
+	// largest bucket is 5 seconds
+	durationMsBuckets = []float64{10, 50, 100, 200, 300, 500, 1000, 2000, 3000, 5000}
+	processName       = filepath.Base(os.Args[0])
+	constLabels       = prometheus.Labels{
+		serviceKey:  processName,
+		instanceKey: getHostname(),
+	}
+	queueMetricLabels = []string{"queue"}
+	taskMetricLabels  = []string{"queue", "type"}
+
+	defTaskCounterOpts = prometheus.CounterOpts{
+		Namespace:   "queue",
+		Subsystem:   "task",
+		Name:        "total_count",
+		Help:        "count of tasks that have been enqueued",
+		ConstLabels: constLabels,
+	}
+	defEnqueueDurationOpts = prometheus.HistogramOpts{
+		Namespace:   "queue",
+		Subsystem:   "task",
+		Name:        "enqueue_duration_ms",
+		Help:        "duration the enqueue action in ms",
+		Buckets:     durationMsBuckets,
+		ConstLabels: constLabels,
+	}
+	defScheduleCounterOpts = prometheus.CounterOpts{
+		Namespace:   "queue",
+		Subsystem:   "scheduler",
+		Name:        "total_scheduled",
+		Help:        "count of tasks that have been scheduled",
+		ConstLabels: constLabels,
+	}
+
+	defSchedulerErrorCounterOpts = prometheus.CounterOpts{
+		Namespace:   "queue",
+		Subsystem:   "scheduler",
+		Name:        "total_errors",
+		Help:        "count of errors while scheduling",
+		ConstLabels: constLabels,
+	}
+
+	// TaskQueueMetrics is the global metrics instance for the task queue of this instance
+	TaskQueueMetrics = TaskQueueMetricsType{
+		Labels: queueMetricLabels,
+		TaskCounter: promauto.NewCounterVec(
+			defTaskCounterOpts,
+			taskMetricLabels,
+		),
+		EnqueueDuration: promauto.NewHistogramVec(
+			defEnqueueDurationOpts,
+			queueMetricLabels,
+		),
+	}
+
+	// SchedulerMetrics is the global metrics instance for the scheduler of this instance
+	SchedulerMetrics = SchedulerMetricsType{
+		Labels: queueMetricLabels,
+		ScheduleCounter: promauto.NewCounterVec(
+			defScheduleCounterOpts,
+			taskMetricLabels,
+		),
+		ErrorCounter: promauto.NewCounterVec(
+			defSchedulerErrorCounterOpts,
+			taskMetricLabels,
+		),
+	}
+)
+
+// SwitchMetricsNamespace changes the namespace used in the metrics, so it can be customized
```

Comment:
```suggestion
// SwitchMetricsNamespace changes the service label used in the metrics, so it can be customized
```

Reply:
good catch, I decided to change the function name as well.